### ISSUE TITLE
Nullish operator doesn't narrowing void type

### ISSUE THREAD

**Issue by Jack-Works** [microsoft/TypeScript#41706]
**TypeScript Version:**  4.2.0-dev.20201124

**Search Terms:**

**Code**

```ts
declare let a: object | void
let b: object
b = a || {}
// Type 'void | object' is not assignable to type 'object'.
//   Type 'void' is not assignable to type 'object'.ts(2322)
b = a ?? {}
```

**Expected behavior:**: No error

**Actual behavior:**: Inconsistent result

**Comment by MartinJohns** [comment]
This is working as intended. Try using the search function to find existing issues.

`void` does not mean no value. It means a value without any type information, a value you should not look at.

**Comment by Jack-Works** [comment]
But if so, why || would work?

**Comment by MartinJohns** [comment]
I don't know about the exact reasons. But generally having a union of `void` and another type is code-smell and makes no sense. `void` means you won't observe the result. This is contradicting having a type whose value you're interested in.

**Comment by Jack-Works** [comment]
If `void` means this reason, I think even `libdom.d.ts` are wrong. Many APIs are known to be returning `undefined` but marked as `void`. This makes a misleading usage of `void`.

**Comment by MartinJohns** [comment]
No, it's fine. `void` means "don't look at the return value, don't care about it". It doesn't matter what these functions return.

**Comment by RyanCavanaugh** [comment]
If you write
```ts
if (console.log("hello")) {
```
that's an obvious logic error. In general if a function *always* returns `undefined`, it's usually better to type it as `void` because it allows these errors to be found more quickly.

**Comment by Jack-Works** [comment]
But why it works on `||`?
